added denovo implementation of da_rnn model

sanskarmodi8 opened the pull request:
Implemented DA-RNN model (LSTM-autoencoder) from wishlist in #6660

fkiraly commented on sktime/forecasting/da_rnn.py:
```diff
@@ -0,0 +1,390 @@
+        params : dict
+            Parameters to create testing instances of the class.
+        """
+        params = {
```
can you include at least two parameter sets?

sanskarmodi8 replied:
Sure!

fkiraly commented on sktime/forecasting/da_rnn.py:
```diff
@@ -0,0 +1,390 @@
+        -------
+        self : reference to self
+        """
+        if not _check_soft_dependencies("torch", severity="none"):
```
you do not need to handle the soft deps here

fkiraly commented on sktime/forecasting/da_rnn.py:
```diff
@@ -0,0 +1,390 @@
+                self.scheduler_.step()
+                epoch_loss += loss.item() * X_batch.size(0)
+            epoch_loss /= len(dataset)
+        self._is_fitted = True
```
you do not need to handle `_is_fitted` here

fkiraly commented on sktime/forecasting/da_rnn.py:
```diff
@@ -0,0 +1,390 @@
+        self.device = device
+        self.random_state = random_state
+
+        self._is_fitted = False
```
you do not need to handle `_is_fitted` here

sanskarmodi8 commented:
Hi @fkiraly ,
I’ve addressed the feedback and updated the code accordingly. Please let me know if any further changes are needed for merging. Looking forward to your thoughts! 
Thanks.

fkiraly reviewed (changes requested):
Nice!

The tests seem to be failing with genuine failures. Have you tried to run `check_estimator` locally on the new classes, to see whether they are also failing locally?

sanskarmodi8 commented:
> Nice!
> 
> The tests seem to be failing with genuine failures. Have you tried to run `check_estimator` locally on the new classes, to see whether they are also failing locally?

Thanks! 

I ran `check_estimator` locally, and some tests are failing for the `DualStageAttentionRNN` class. 

However, I'm not sure how to resolve them. Could you provide some guidance on debugging or point me to any similar examples?

sanskarmodi8 commented:
Hey @fkiraly,

I have updated the implementation, and now the `check_estimator` passes all tests on my local.

![image](https://github.com/user-attachments/assets/8b18a7b8-9c86-486a-808a-2aef7c55a3fb)

fkiraly reviewed (changes requested):
Look good and ready for merge, some requests relate to tdocumentation:

* we need to add the estimator the API reference for forecasting - simply follow the pattern there (`forecasting.rst`)
* in the docstring, we should explain the NN structure a bit better - what is the structure of the neural network?

sanskarmodi8 commented:
> Look good and ready for merge, some requests relate to tdocumentation:
> 
> * we need to add the estimator the API reference for forecasting - simply follow the pattern there (`forecasting.rst`)
> * in the docstring, we should explain the NN structure a bit better - what is the structure of the neural network?

I have updated the code accordingly.

fkiraly commented on docs/source/api_reference/forecasting.rst:
```diff
@@ -463,6 +463,14 @@ Structural time series models
+    :toctree: auto_generated/
+    :template: class.rst
+
+    DualStageAttentionRNN
```
Could you move this after `CINNForecaster`? I think that fits thematically better.

sanskarmodi8 replied:
sure
